keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+  cumbersome.
+* The same issues apply when doing changes to the config of a presubmit
+* When a project maintains multiple branches, e.G. because there are release branches, the
+  maintainers must remember to create a copy of the presubmit in the `infra-config` repository.
+  Otherwise it is easely possible that the presubmit becomes incompatible with one of the branches
+  it is used in, because somone makes a change to its config and forgets to test against all branches.
+  This is an additional step maintainers have to remember when branching off.
+
+
+### Goals
+
+* It is possible to version some or all presubmits of a given repository inside that repository in a
+  `yaml` file
+* This feature is optional and opt-in
+* The triggering of presubmits on pull request creation or updates continues to work and includes the
+  jobs that are managed inside the code repository
+* Re-Running tests via the `/retest` command continues to work and includes the jobs that are
+  managed inside the code repository
+* Explicitly executing optional tests via the `/test <<myjob>>` command continues to work and includes
+  all jobs that are managed inside the code repository
+* All the existing defaulting and validation for Presubmit jobs is being used to default and validate
+  jobs that are managed inside the code repository
+* If there is an error parsing, defaulting or validating the presubmits that are managed inside the
+  code repository, a comment will be posted to GitHub stating the error
+* Pull Requests on which an error occurred during parsing, defaulting or validation the presubmits that
+  are managed in the code repository are not consideres as merge candidates by Tide
+* Tide executes the presubmits that are defined inside the code repository when it re-tests
+
+### Non-Goals
+
+* The option to configure Postsubmits or Periodics inside the tested repository. This may be
+  done in a future iteration.
+
+## Proposal
+
+It is proposed to introduce a the option to configure additional presubmits inside the code
+repositories that are tested by prow via a file named `prow.yaml`
+
+The minimum part of Prow this has to be integrated with is:
+
+* The `trigger` plugin as it manages the triggering of Jobs upon Pull Request Events or commands
+* `Tide`, as it re-excutes jobs when the base branch for a pull request changed after it was
+  tested
+
+The following changes will be required:
+
+#### `prow/config` adjustments
+
+The `prow/config` package needs to unite its various defaulting and
+validation funcs for presubmits so that in the end a func can be
+exported that will be used both in case of config reload and
+when additional presubmits are loaded from a `prow.yaml`.
+
+*Open question:* What would be the best signature for such a function?
+It must be able to optionally accept additional presubmits for one repo
+
+#### New `prow/inrepoconfig/api` package
```

Comment:
This is pretty low-level for a KEP IMO

Reply:
I agree in that this is pretty low level for a kep but I was asked for a design so I made this a bit more detailled. It may not matter so much where new packages exactly are, but we should agree on the question if we actually create a new package or put the new functionality in existing ones